Delegate toXContent logic from ClusterState to its member classes #48218

zacharymorn opened the pull request:
ClusterState.toXContent contains duplicate logic from its member classes such as IndexMetaData.toXContent and MetaData.toXContent etc. This PR is to remove the duplicate and delegate logic to those classes. For details, please refer to #48218

zacharymorn committed:
Add temporary XContent comparison test to detect potential changes to serialization for IndexMetaData.
This test may sometimes fail due to JsonNode.equals may be order sensitive for deeply nested json structure

zacharymorn committed:
Subtitute relevant logic in ClusterState.toXContent with IndexMetaData.toXContent, and note the difference both in code as comment and in serialization.
This test may sometimes fail due to JsonNode.equals may be order sensitive for deeply nested json structure

zacharymorn committed:
Add temporary XContent comparison test to detect potential changes to serialization for IndexTemplateMetaData
This test may sometimes fail due to JsonNode.equals may be order sensitive for deeply nested json structure

zacharymorn committed:
Subtitute relevant logic in ClusterState.toXContent with IndexTemplateMetaData.toXContent, and note the difference in serialization.
This test may sometimes fail due to JsonNode.equals may be order sensitive for deeply nested json structure

zacharymorn committed:
Subtitute relevant logic in ClusterState.toXContent with MetaData.toXContent, and note the difference in serialization.
This test may sometimes fail due to JsonNode.equals may be order sensitive for deeply nested json structure

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Distributed)

zacharymorn committed:
Update tests to verify invocation of member class toXContent, instead of testing serializaation directly

zacharymorn committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into issue-48218

gwbrown commented:
Hi @zacharymorn, first off I want to apologize for taking so long to review this.  I've been waiting on some internal discussions that took a while to happen.  I want to thank you especially for looking into the other places in ClusterState's toXContent that have this issue.

We discussed this recently in the Distributed team sync, and we're going to:
1) Move the list of other duplicated toXContent logic to a meta-issue so that we can keep track of which ones we've addressed. This probably means we'll want to remove the TODO comments before merging (but leave them for the moment, until we get them transferred to a meta-issue).
2) Review the changes to the Cluster State JSON response, as we may want to make changes to make it more easily human-readable (as this JSON is mostly used as a toubleshooting tool).
3) Do some testing with other applications (for example, Kibana) to make sure we aren't going to accidentally break something if we backport this to 7.x. While we [document that the format of this API may change in minor versions](https://www.elastic.co/guide/en/elasticsearch/reference/7.5/cluster-state.html), we still don't want to do so casually and risk breakage.

We'll also review the code changes in this PR as usual soon.

ywelsch commented on server/src/test/java/org/elasticsearch/action/admin/cluster/reroute/ClusterRerouteResponseTests.java:
```diff
@@ -82,8 +82,10 @@ public void testToXContent() throws IOException {
                     "      }\n" +
                     "    },\n" +
-                    "    \"metadata\" : {\n" +
+                    "    \"meta-data\" : {\n" +
```
If anyone was relying on the stability of the cluster state APIs, this is definitely the biggest thing that would break them. I wonder if we should change `MetaData.toXContent` to have metadata there, and be lenient in the `MetaData.fromXContent` method, allowing both. This means that the internal storage format can handle the change, and the user-facing cluster state will be kept the same as well.

ywelsch replied:
An alternative would be to use `meta-data` / `metadata` in `MetaData.toXContent` depending on the context (API vs GATEWAY).

zacharymorn replied:
Yes I think this makes sense. Will try to parametrize the code.

ywelsch commented on server/src/test/java/org/elasticsearch/action/admin/cluster/reroute/ClusterRerouteResponseTests.java:
```diff
@@ -93,28 +95,28 @@ public void testToXContent() throws IOException {
-                    "          \"aliases\" : [ ],\n" +
-                    "          \"primary_terms\" : {\n" +
-                    "            \"0\" : 0\n" +
+                    "            \"index.max_script_fields\" : \"10\",\n" +
```
MetaData.toXContent could have `flat_settings` defaulting to `false` if the context is API. That means that the representation would stay the same here.

ywelsch commented on server/src/test/java/org/elasticsearch/action/admin/cluster/reroute/ClusterRerouteResponseTests.java:
```diff
@@ -93,28 +95,28 @@ public void testToXContent() throws IOException {
+                    "            \"index.number_of_replicas\" : \"0\",\n" +
+                    "            \"index.number_of_shards\" : \"1\",\n" +
+                    "            \"index.shard.check_on_startup\" : \"true\",\n" +
+                    "            \"index.version.created\" : \"8000099\"\n" +
```
the version is hardcoded here, which means that this will break as  soon as we do  a version bump. Please use a variable as it used to be before.

zacharymorn replied:
Will do.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/ClusterState.java:
```diff
@@ -419,106 +414,14 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
-            coordinationMetaData().toXContent(builder, params);
-            builder.endObject();
-
-            builder.startObject("templates");
```
Would be good to have an example that shows how / whether templates are rendered differently now.

zacharymorn replied:
I have it here as an intermediate commit https://github.com/elastic/elasticsearch/pull/48924/commits/bfc1103163f4cb1fb6c49d5a3bff06e50a30eef6 . Please let me know if this look good.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/ClusterState.java:
```diff
@@ -419,106 +414,14 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
-            builder.endObject();
-
-            builder.startObject("indices");
-            for (IndexMetaData indexMetaData : metaData()) {
```
With this change, we are now also exposing index metadata customs. I'm not sure if these might contain sensitive information. Need to double-check. We could perhaps omit these for API context.

zacharymorn replied:
I think I'm missing some knowledge about the potential sensitive information here in metadata custom. Let me see if I can find some documentation about it.

ywelsch commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -1205,6 +1211,23 @@ public static void toXContent(IndexMetaData indexMetaData, XContentBuilder build
             builder.endArray();
 
+/*
+            ==== for the above, ClusterState.toXContent has the following ====
```
sure, but how is this rendered differently? Can we see an example?

zacharymorn replied:
Yes it's available here as an intermediate commit https://github.com/elastic/elasticsearch/pull/48924/commits/4850accccfb0a9d0c8cc810cb6813ec07e1e8e47 . Please let me know if this looks good.

ywelsch reviewed (commented):
I've left some initial comments. I would prefer to start by parameterizing the code in `MetaData.toXContent` to, depending on the context (API vs GATEWAY), emit different XContent. We can then step-by-step resolve the differences between the two. Also we need more detailed examples to see how the output between the two differs. A good first test would be to take a complex cluster  state, use the old serialization method, and check that the new parameterized one in `MetaData.toXContent` provides equivalent output.

zacharymorn commented:
@gwbrown No problem and I had some delay on my side as well. The caution taken definitely makes sense as I too also see changes in serialization that could potentially break clients. Please let me know when you guys have detailed plans for #1 and #3.

zacharymorn commented:
@ywelsch Yes I agree with the approach of seeing the changes in serialization as well. I did start my changes by generating an existing serialization of the cluster state (https://github.com/elastic/elasticsearch/pull/48924/commits/df24188f1ee832f878f03807347dee51424044ef), and each time I updated a portion of code I committed both the changes to code and to the serialization. In my last commit though, I removed the checked for serialization content and replaced it with a verification of delegating call (to member classes). 

I think the problem here is, as the existing serialization of a complex cluster state are new changes in my existing PR, it might be hard to use git to clearly shows the changes of serialization introduced by the PR globally, as they will always be viewed as new changes. Will it help if I create a separate PR to just check in the existing serialization, and track the updates to the serialization in this PR?

ywelsch commented:
yes, this PR is trying to do too many things at once. Let's start by having a PR first that parameterizes the code in `MetaData.toXContent` to, depending on the context (API vs GATEWAY), emit different XContent, providing further tests that the serialization between new and old method is identical. In follow-ups, we can then start removing the old logic, and make adjustments to bring API and GATEWAY serialization closer.

zacharymorn commented:
@ywelsch As this current PR has too many commits to cleanly show the changes, I’ve made a new branch and PR https://github.com/elastic/elasticsearch/pull/52743  to better track the changes in the new direction. I'll close this PR for now.

As MetaData.toXContent is already parameterized as it checks for API context for certain serializations,  I’ve organized the 4 commits in that PR as follows:
1. Add API and Gateway serialization tests for complex ClusterState and MetaData objects without modifying production code
2. Copy over ClusterState.toXContent serialization test metadata section to  MetaData.toXContent API serialization tests to highlight the difference and setup baseline 
3. Parameterize MetaData.toXContent further and have its serialization under API context to be as close as possible to ClusterState. 
	1. Note that as the updates to existing serialization test cases show, there are inconsistencies of serializations under API context.
	2. We can pick and choose which serializaton scheme we actually want, and which one should no longer be supported under API context.
4. Replace ClusterState.toXContent metadata section with parameterized MetaData.toXContent
